### ISSUE TITLE
Take care of duplicate edges and loops

### ISSUE THREAD

**Issue by khosravipasha** [waynebhayes/SANA#5]
In the Graph might have dupliate edges.

The adjacentyList will have duplicate edges, we want to not add it twice. We can check the adjacency matrix and not add duplicate edges.

Also need to take care of self-loops. # #4

**Comment by lckrdg** [comment]
So do we only want to eliminate duplicate edges from the AdjMatrix?

**Comment by khosravipasha** [comment]
The AdjMatrix would be ok, because it would be overwritten. Bur for adjList we would have the edge twice. For example, if we have edge "1 2" twice. Then the adjList becomes.
[1] -> [2, 2]
[2] -> [1.1]

But the one we want is just
[1] -> [2]
[2] -> [1]

**Comment by waynebhayes** [comment]
How hard would it be to save half the memory in adjacency lists and only
store the edge in the n1 < n2 list?  So then a query on the adjacency list
for (u,v), we'd need to swap u and v if u > v?

-- 
Wayne Hayes, Ph.D.
Associate Professor, Computer Science, University of California, Irvine
President Trump is a self-absorbed, self-serving, habitual liar. The former
two are opinion; the latter is fact. Examples (1)
http://foreignpolicy.com/2017/08/18/trumps-lies-about-james-comey-keep-unraveling
. (2) https://www.theguardian.com/us-news/trump-administration

**Comment by lckrdg** [comment]
Yeah, should be easy enough. I'm going to modify the AddEdge method in Graph.cpp to reflect this change.

**Comment by red4711** [comment]
Is this issue closed?

**Comment by lckrdg** [comment]
I think so, development doesn't ever insert high to low edges. It only inserts low to high.

from lib-sana/graph/graph.cpp
```
 if (node1 > node2) {
    if (adjLists[node2][node1] == 0) {
        adjLists[node2].push_back(node1);
    }
} else {
    adjLists[node1].push_back(node2);
}

```
We need to adjust the lookup still I believe, so if we're trying to find edges, we need to always lookup low to high.

**Comment by waynebhayes** [comment]
Hey, Pasha! I may be visiting the Bay Area in the next few months. Let me
know if you get this message and we can connect.

**Comment by khosravipasha** [comment]
Good idea, sent you an email :)
